### PR TITLE
Add DistributedDataParallel

### PR DESCRIPTION
**Pull request by apaszke**
```python
import torch
import torch.distributed as dist
import torchvision

dist.init_process_group(backend='gloo')

model = torchvision.models.resnet50().cuda()
model = torch.nn.DistributedDataParallel(model) # prepend Distributed

dataset = ...
# so each process sees only a subset of the whole dataset
sampler = torch.utils.data.DistributedSampler(dataset)
data_loader = torch.utils.data.DataLoader(
        ..., sampler=sampler)

for batch, target in data_loader:
    # optimize model, log, etc.
```

### CODE REVIEW

File: torch/lib/THD/base/data_channels/DataChannelMPI.cpp [hunk 1]
```diff
@@ -476,7 +476,7 @@ THDGroup DataChannelMPI::newGroup(const std::vector<rank_type>& ranks) {
   MPI_Group_incl(world_group, int_ranks.size(), int_ranks.data(), &ranks_group);
 
   MPI_Comm new_comm;
-  MPI_Comm_create_group(MPI_COMM_WORLD, ranks_group, 0, &new_comm);
+  //MPI_Comm_create_group(MPI_COMM_WORLD, ranks_group, 0, &new_comm);
```
Comment (VirrageS) [inline torch/lib/THD/base/data_channels/DataChannelMPI.cpp]: I do not think this should be commented 😄 
```
MPI_Comm_create(MPI_COMM_WORLD, ranks_group, &new_comm);
```

File: torch/lib/THD/base/Cuda.hpp [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+extern THCState** _THDCudaState;
+
+inline THCState* THDGetCudaState() {
+  return *_THDCudaState;
+}
+
+int THDGetStreamId(cudaStream_t stream);
+
+#include "Cuda.h"
```
Comment (colesbury) [inline torch/lib/THD/base/Cuda.hpp]: Why is the include at the end of the file?

File: torch/lib/THD/base/init_methods/InitMethodTCP.cpp [hunk 3]
```diff
@@ -35,8 +35,11 @@ std::string getRandomString()
 
   int fd;
   unsigned int seed;
+  ssize_t bytes_read;
   SYSCHECK(fd = open("/dev/urandom", O_RDONLY));
-  SYSCHECK(read(fd, &seed, sizeof(seed)));
+  SYSCHECK(bytes_read = read(fd, &seed, sizeof(seed)));
```
Comment (colesbury) [inline torch/lib/THD/base/init_methods/InitMethodTCP.cpp]: Why do you need to read from `/dev/urandom` (instead of just using the C++11 random library)?

File: torch/nn/parallel/distributed.py [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+        system in every iteration.
+
+    Args:
+        module: module to be parallelized
+        device_ids: CUDA devices (default: all devices)
+        output_device: device location of output (default: device_ids[0])
+
+    Example::
+
+        >>> net = torch.nn.DataParallel(model, device_ids=[0, 1, 2])
```
Comment (colesbury) [inline torch/nn/parallel/distributed.py]: `s/DataParallel/DistributedDataParallel`

If there are environment variables that need to be set to use it, this would be a good place give some examples of that too.

Same with any functions you need to call to rendezvous before creating DistributedDataParallel

File: torch/utils/data/distributed.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import math
+import torch
+from .sampler import Sampler
+
+
+class DistributedSampler(Sampler):
```
Comment (colesbury) [inline torch/utils/data/distributed.py]: Class level docs here would be good

**Review by colesbury — COMMENTED** [review]
The parts I understand look good. A few questions about things that confused me:

File: torch/lib/THD/base/Cuda.cpp [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#include "Cuda.hpp"
+
+THCState** _THDCudaState;
```
Comment (apaszke) [inline torch/lib/THD/base/Cuda.cpp]: It just makes it possible to lazily initialize THC, even after THD. Instead of giving a `THCState*` (which must be valid) to distributed, you can give it a pointer to the place where `THCState*` will be initialized. You're guaranteed that it will be properly set by the time you get a first CUDA tensor in THD.

File: torch/lib/THD/base/Cuda.hpp (hunk 2 above)
Comment (apaszke) [inline torch/lib/THD/base/Cuda.hpp]: It's a header that's included in other files. Functions form `.h` should also be exposed here

File: torch/lib/THD/base/init_methods/InitMethodTCP.cpp (hunk 3 above)
Comment (apaszke) [inline torch/lib/THD/base/init_methods/InitMethodTCP.cpp]: Because C++11 random library either doesn't guarantee anything about the seed, or uses a fixed one, while we need to somehow break symmetry between machines.

Comment (colesbury) [inline torch/lib/THD/base/init_methods/InitMethodTCP.cpp]: Can you use http://en.cppreference.com/w/cpp/numeric/random/random_device ?

File: torch/lib/THD/base/Cuda.cpp (hunk 6 above)
Comment (colesbury) [inline torch/lib/THD/base/Cuda.cpp]: That seems janky. Can't you just initialize THD after constructing the THCState*?

Comment (apaszke) [inline torch/lib/THD/base/Cuda.cpp]: It's a bit ugly because now I need to insert code to a CUDA module that will initialize it. But I can do it, if you really don't like it

File: torch/lib/THD/base/Cuda.hpp (hunk 2 above)
Comment (colesbury) [inline torch/lib/THD/base/Cuda.hpp]: Can it be included at the top of the file?

File: torch/lib/THD/base/init_methods/InitMethodTCP.cpp (hunk 3 above)
Comment (apaszke) [inline torch/lib/THD/base/init_methods/InitMethodTCP.cpp]: Second sentence:

> std::random_device may be implemented in terms of an implementation-defined pseudo-random number engine if a non-deterministic source (e.g. a hardware device) is not available to the implementation. In this case each std::random_device object may generate the same number sequence.

Comment (colesbury) [inline torch/lib/THD/base/init_methods/InitMethodTCP.cpp]: I forgot how messed up the C++11 random library is.

What's the point of the `uid` here? It seems like you already have uniqueness with the group_name and rank.

If that's not sufficient, then why not just use the single value you get from `/dev/urandom`? There's no additional randomness (or uniqueness) from passing it through the Mersenne twister or the encoding it into the vocabulary.

Comment (apaszke) [inline torch/lib/THD/base/init_methods/InitMethodTCP.cpp]: `rank` don't necessarily have to be specified, so you need these uids to perform automatic rank assignment.

You're right, I'll simplify the code tomorrow.

File: torch/nn/parallel/distributed.py (hunk 4 above)
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: The only function needed to call is distributed initialization, which only requires any env vars if you specify an env init method. It will raise an error if you don't do this first.

Comment (soumith) [inline torch/nn/parallel/distributed.py]: I think sam's point was that this should be a fuller example that users can see and immediately understand how to use

**Comment by apaszke** [comment]
Pushed fixes for review comments in a separate commit. I will reset it and squash the changes into appropriate commits when they're accepted.

**Review by colesbury — APPROVED** [review]

File: test/run_test.sh [hunk 7]
```diff
@@ -20,72 +20,37 @@ fi
 
 pushd "$(dirname "$0")"
 
-echo "Running torch tests"
-$PYCMD test_torch.py $@
```
Comment (fmassa) [inline test/run_test.sh]: Is it intended to remove those tests here?

Comment (apaszke) [inline test/run_test.sh]: uh no!

**Comment by acgtyrant** [comment]
@apaszke Hi, as far as I understand, the `DistributedDataParallel` `nccl.reduce` all gradients from all machines to the machine whose rank is 0(device to device), and every machine `comm.reduce` all gradients from its GPUs to the device 0 respectively again. After every machine updates the module in device 0 respectively, the module in device 0 will `_sync_params` the parameters to others devices respectively in the forward function. Right?

I am suprised that the `DistributedDataParallel` does not broadcast the parameters from one machine to all machines like DataParallel broadcasting the parameters from device 0 to all devices, I think this means that the traning modules in different machines are not same.

**Comment by soumith** [comment]
@acgtyrant DistributedDataParallel has only one synchronization, all gradients are all_reduce to all machines, so all machines have same copy of gradients. Then every machine does it's own optimization step.
